### PR TITLE
Created SlashCommands

### PR DESCRIPTION
**Pull request by alexyy802**
## Summary

Slash Commands

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [x] This PR adds something new (e.g. new method or parameters).

### CODE REVIEW

File: discord/ext/commands/slash.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
+import discord
+import aiohttp
+import inspect
+from .. import commands
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:4]: Unused import
```suggestion
```

File: discord/ext/commands/slash.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+            if annotation is parameter_type:
+                t = discord_value
+
+        options.append(
+            {
+                "name": k,
+                "description": k,
+                "type": t,
+                "required": required,
+                "choices": command_choices,
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:58]: what is "choices" for?

File: discord/ext/commands/slash.py [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+async def _delete(url, json_dict=None, headers=None):
+    async with aiohttp.ClientSession() as session:
+        async with session.delete(url, json=json_dict, headers=headers) as response:
+            try:
+                return json.loads(await response.text())
+            except json.decoder.JSONDecodeError:
+                return await response.text()
+
+
+def _get_headers(client):
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:101]: isn't it Client not client

File: discord/ext/commands/slash.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+        self.pinned = discord_dict.get("pinned")
+        self.mention_everyone = discord_dict.get("mention_everyone")
+        self.tts = discord_dict.get("tts")
+        self.created_at = datetime.datetime.now()
+        self.edited_timestamp = discord_dict.get("edited_timestamp")
+        self.flags = discord_dict.get("flags")
+
+    async def edit(self, content):
+        await _patch(
+            f"https://discord.com/api/v9/webhooks/{self.client.user.id}/{self.token}/messages/@original",
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:212]: V9? we are only at V8 iirc

File: discord/ext/commands/slash.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+            headers=_get_headers(self.client),
+            json_dict={"content": content},
+        )
+
+    async def delete(self):
+        await _delete(
+            f"https://discord.com/api/v9/webhooks/{self.client.user.id}/{self.token}/messages/@original",
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:219]: Again V9

File: discord/ext/commands/slash.py [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+        }
+
+
+def _add_commands(bot_commands, command_list, choices, hidden, client):
+    for x in bot_commands:
+        command_list.append(_create_info(x, choices))
+
+    for x in command_list:
+        w = _post_sync(
+            f"https://discord.com/api/v9/applications/{client.user.id}/commands",
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:292]: V9 again

File: discord/ext/commands/slash.py [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+        )
+        try:
+            if w["name"] == x["name"]:
+                print(f"Synced command {x['name']}")
+        except KeyError:
+            print(f"Error syncing command {x['name']}: {w}")
+        time.sleep(10)
+
+    slash_commands = _get_sync(
+        f"https://discord.com/api/v9/applications/{client.user.id}/commands",
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:304]: V9 again

File: discord/ext/commands/slash.py [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+        headers=_get_headers(client),
+    )
+    name_list = []
+    for x in command_list:
+        name_list.append(x["name"])
+    for x in slash_commands:
+        if x["name"] not in name_list or x["name"] in hidden:
+            t = _get_sync(
+                f"https://discord.com/api/v9/applications/{client.user.id}/commands/{x['id']}",
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:313]: V9 again

File: discord/ext/commands/slash.py [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+                headers=_get_headers(client),
+            )
+            if t == "":
+                print(f"Removed command {x['name']}")
+
+
+async def sync_all_commands(
+    client: typing.Union[
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:321]:
```suggestion
    Client: typing.Union[
```

File: discord/ext/commands/slash.py [hunk 10]
```diff
@@ -0,0 +1,534 @@
...
+        flags = None
+
+    async def on_socket_response(msg):
+        if msg["t"] == "INTERACTION_CREATE":
+
+            args = []
+            kwargs = {}
+
+            ack = await _post(
+                f"https://discord.com/api/v9/interactions/{msg['d']['id']}/{msg['d']['token']}/callback",
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:361]: V9 Again

File: discord/ext/commands/slash.py [hunk 11]
```diff
@@ -0,0 +1,534 @@
...
+                },
+            )
+
+            for x in client.commands:
+                try:
+                    msg["d"]["data"]["name"]
+                except KeyError:
+                    for functions in button_functions:
+                        info = msg["d"]["data"]["custom_id"].split(" ")
+                        message_components = msg["d"]["message"]["components"][0][
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:375]: Components? what is this doing here exactly?

File: discord/ext/commands/slash.py [hunk 12]
```diff
@@ -0,0 +1,534 @@
...
+
+                        await x(*args, **kwargs)
+
+                    except Exception as error:
+                        if error_function != None:
+                            await error_function(SlashContext(msg, client), error)
+                        if error_function == None:
+                            await args[0].send(str(error))
+
+    client.add_listener(on_socket_response)
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:527]: on_socket_response was removed in V2 this will **Not** work

File: discord/ext/commands/slash.py [hunk 13]
```diff
@@ -0,0 +1,534 @@
...
+
+
+BUTTON_PRIMARY = 1
+BUTTON_SECONDARY = 2
+BUTTON_SUCCESS = 3
+BUTTON_DANGER = 4
+BUTTON_LINK = 5
```
Comment (VincentRPS) [inline discord/ext/commands/slash.py:534]: I don't know why this is needed?

**Review by VincentRPS — CHANGES_REQUESTED** [review]
Weird stuff and all just fix em

File: discord/ext/commands/slash.py (hunk 2 above)
Comment (alexyy802) [reply discord/ext/commands/slash.py:58]: the dropdowns i think

Comment (VincentRPS) [reply discord/ext/commands/slash.py:58]: @alexyy802 Okay if you say so

File: discord/ext/commands/slash.py (hunk 11 above)
Comment (alexyy802) [reply discord/ext/commands/slash.py:375]: it supports buttons as well that's why it has components

File: discord/ext/commands/slash.py (hunk 13 above)
Comment (alexyy802) [reply discord/ext/commands/slash.py:534]: For button types

Comment (VincentRPS) [reply discord/ext/commands/slash.py:534]: Just state them when a user makes a button
@alexyy802

**Pull request merged**